### PR TITLE
print msg when png or layout takes a long time

### LINKED ISSUE #271: cli: Print messages when layout is taking while to let the user know we're not frozen
See discussion in https://github.com/terrastruct/d2/pull/224

### LINKED ISSUE #915: add progress spinner to the CLI
there's 3 places that can take noticeable time:
1. layout
2. fetching images
3. one-time download for playwright dependencies for png

we already have a "...still working" outputting periodically in stage 2, but we don't have anything for stage 1 or 3.

### LINKED ISSUE #1057: need progress cli when png converting
Screen Shot 2023-03-17 at 10 42 19 PM

### PR DESCRIPTION
**Pull request by alixander**
Screen Shot 2023-03-18 at 1 03 33 AM

closes #1057
closes #271
closes #915

### CODE REVIEW

**Comment by alixander** [comment]
nvm that review, that doesn't work. goddamnit why is go concurrency so complicated: https://stackoverflow.com/questions/17797754/ticker-stop-behaviour-in-golang

File: d2cli/main.go [hunk 1]
```diff
@@ -285,11 +285,33 @@ func compile(ctx context.Context, ms *xmain.State, plugin d2plugin.Plugin, sketc
 	if sketch {
 		opts.FontFamily = go2.Pointer(d2fonts.HandDrawn)
 	}
+
+	t := time.NewTicker(time.Second * 5)
+	defer t.Stop()
+	stop := make(chan struct{}, 1)
+	defer func() {
```
Comment (nhooyr) [inline d2cli/main.go:292]:
```suggestion
	defer close(stop)
```

File: lib/png/png.go [hunk 2]
```diff
@@ -82,6 +83,24 @@ var genPNGScript string
 const pngPrefix = "data:image/png;base64,"
 
 func ConvertSVG(ms *xmain.State, page playwright.Page, svg []byte) ([]byte, error) {
+	t := time.NewTicker(time.Second * 5)
```
Comment (nhooyr) [inline lib/png/png.go:86]: Instead of duplicating all this code can you move into `lib` and reuse?

File: d2cli/main.go [hunk 3]
```diff
@@ -285,11 +285,33 @@ func compile(ctx context.Context, ms *xmain.State, plugin d2plugin.Plugin, sketc
...
+		stop <- struct{}{}
+	}()
+
+	go func() {
+		for {
+			select {
+			case <-t.C:
+				ms.Log.Info.Printf("compiling & running layout algorithms...")
+			case <-stop:
+				ms.Log.Info.Printf("done")
```
Comment (nhooyr) [inline d2cli/main.go:302]: Is this necessary?

**Review by nhooyr — CHANGES_REQUESTED** [review]

File: d2cli/main.go (hunk 3 above)
Comment (alixander) [reply d2cli/main.go]: oops, thx for catching, it was just to debug that it actually stops

File: lib/background/background.go [hunk 4]
```diff
@@ -0,0 +1,28 @@
+package background
+
+import "time"
+
+func DoEveryX(do func(), x time.Duration) (kill func()) {
+	stopped := false
+	t := time.NewTicker(x)
+	stop := make(chan struct{})
```
Comment (nhooyr) [inline lib/background/background.go:8]:
```suggestion
	done := make(chan struct{})
```

File: lib/background/background.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package background
+
+import "time"
+
+func DoEveryX(do func(), x time.Duration) (kill func()) {
+	stopped := false
```
Comment (nhooyr) [inline lib/background/background.go:6]: move this above the callback below where it's used.

File: lib/background/background.go [hunk 6]
```diff
@@ -0,0 +1,28 @@
+package background
+
+import "time"
+
+func DoEveryX(do func(), x time.Duration) (kill func()) {
```
Comment (nhooyr) [inline lib/background/background.go:5]:
```suggestion
func Repeat(do func(), x time.Duration) (cancel func()) {
```

File: lib/background/background.go [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+			case <-stop:
+				return
+			}
+		}
+	}()
+
+	return func() {
+		if !stopped {
+			stopped = true
+			t.Stop()
```
Comment (nhooyr) [inline lib/background/background.go:24]: defer this inside the go func instead of doing it here.

File: lib/png/png.go [hunk 8]
```diff
@@ -82,6 +84,12 @@ var genPNGScript string
 const pngPrefix = "data:image/png;base64,"
 
 func ConvertSVG(ms *xmain.State, page playwright.Page, svg []byte) ([]byte, error) {
+	kill := background.DoEveryX(func() {
+		ms.Log.Info.Printf("converting to PNG...")
+	}, time.Second*5)
+
```
Comment (nhooyr) [inline lib/png/png.go:90]:
```suggestion
```

**Review by nhooyr — CHANGES_REQUESTED** [review]

**Review by nhooyr — APPROVED** [review]
✅

**Pull request merged**